nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1051 @@
+                        );
+
+                        // break here - it's possible that the running snapshot
+                        // record was GCed, and it won't come back.
+                        Ok(())
+                    }
+                }
+            },
+            |e: WaitError, delay| {
+                info!(
+                    log,
+                    "{:?}, trying again in {:?}",
+                    e,
+                    delay;
+                    "region_id" => %region_id,
+                    "snapshot_id" => %snapshot_id,
+                );
+            }
+        )
+        .await
+        .map_err(|e| match e {
+            WaitError::Transient(e) => {
+                // The backoff crate can be configured with a maximum elapsed time
+                // before giving up, which means that Transient could be returned
+                // here. Our current policies do **not** set this though.
+                Error::internal_error(&e.to_string())
+            }
+
+            WaitError::Permanent(e) => {
+                e
+            }
+        })
+    }
+
+    pub async fn delete_crucible_snapshot(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(), Error> {
+        self.delete_crucible_snapshot_impl(log, dataset, region_id, snapshot_id)
+            .await
+    }
+
+    async fn delete_crucible_snapshot_impl(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(), Error> {
+        // Unlike other Crucible agent endpoints, this one is synchronous in that it
+        // is not only a request to the Crucible agent: `zfs destroy` is performed
+        // right away. However this is still a request to illumos that may not take
+        // effect right away. Wait until the snapshot no longer appears in the list
+        // of region snapshots, meaning it was not returned from `zfs list`.
+
+        let dataset_id = dataset.id();
+
+        info!(
+            log,
+            "requesting region snapshot delete";
+            "dataset_id" => %dataset_id,
+            "region_id" => %region_id,
+            "snapshot_id" => %snapshot_id,
+        );
+
+        self.request_crucible_snapshot_delete(
+            log,
+            dataset,
+            region_id,
+            snapshot_id,
+        )
+        .await?;
+
+        backoff::retry_notify(
+            backoff::retry_policy_internal_service_aggressive(),
+            || async {
+                let response = match self
+                    .get_crucible_region_snapshots(log, dataset, region_id)
+                    .await
+                {
+                    Ok(v) => Ok(v),
+
+                    // Return Ok if the dataset's agent is gone, no
+                    // delete call is required.
+                    Err(Error::Gone) => return Ok(()),
```

Comment:
I finally noticed here, that we don't log when something is gone. Should we add that logging to the `ProgenitorOperationRetry` code or leave it to the users. In the latter case, we should add a log here, as well as all other places `Gone` is returned.

Reply:
Good point - I don't think `ProgenitorOperationRetry` is a good place for this because it doesn't have any visibility into _what_ is gone, just that the gone_check function returned true, so I've added it one layer up: see 0baa9c8e5

Reply:
Great! Thanks.